File: frontend/src/api/edit.ts
import http from '@/api/http'

interface ApiResponse<T> {
  code: string
  message: string
  data: T
  trace_id: string
}

export type EditTool = 'polygon' | 'line_buffer' | 'brush_path'
export type EditVariable = 'qpf' | 'ptype'
export type EditOperation =
  | 'set_value'
  | 'increase'
  | 'decrease'
  | 'multiply'
  | 'clear'
  | 'ptype_set'
  | 'set_ptype'
  | 'screen_clear'

export interface EditPreviewRequest {
  session_id: string
  tool: EditTool
  variable: EditVariable
  operation: EditOperation
  mask: Record<string, unknown>
  parameters: Record<string, unknown>
}

export interface EditPreviewResponse {
  preview_id: string
  affected_grid_count: number
  affected_area_km2: number
  area_mode?: string
  before_stats: Record<string, number>
  after_stats: Record<string, number>
  op_ptype_transition: Record<string, number> | null
  new_precip_needs_ptype: boolean
  new_precip_count: number
  warnings: Array<{ code: string; count: number }>
  preview_image?: string | null
}

export interface EditApplyRequest {
  session_id: string
  preview_id: string
  target_ptype?: number
}

export interface EditApplyResponse {
  operation_id: string
  sequence_no: number
  applied: boolean
  can_undo: boolean
  can_redo: boolean
}

export interface UndoRedoRequest {
  session_id: string
}

export interface UndoRedoResponse {
  can_undo: boolean
  can_redo: boolean
  operation_count: number
}

export interface OperationItem {
  sequence_no: number
  tool_name: string
  operation_type: string
  variable_name: string
  affected_grid_count: number
  is_undone: number
  created_at: string
}

export interface OperationListResponse {
  operations: OperationItem[]
}

export async function editPreview(params: EditPreviewRequest): Promise<EditPreviewResponse> {
  const { data } = await http.post<ApiResponse<EditPreviewResponse>>('/edit/preview', params)
  return data.data
}

export async function editApply(params: EditApplyRequest): Promise<EditApplyResponse> {
  const { data } = await http.post<ApiResponse<EditApplyResponse>>('/edit/apply', params)
  return data.data
}

export async function editUndo(params: UndoRedoRequest): Promise<UndoRedoResponse> {
  const { data } = await http.post<ApiResponse<UndoRedoResponse>>('/edit/undo', params)
  return data.data
}

export async function editRedo(params: UndoRedoRequest): Promise<UndoRedoResponse> {
  const { data } = await http.post<ApiResponse<UndoRedoResponse>>('/edit/redo', params)
  return data.data
}

export async function getOperations(sessionId: string): Promise<OperationListResponse> {
  const { data } = await http.get<ApiResponse<OperationListResponse>>('/edit/operations', {
    params: { session_id: sessionId },
  })
  return data.data
}

File: backend/app/schemas/edit.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict, Field, model_validator


ToolName = Literal["polygon", "line_buffer", "brush_path", "lasso"]
VariableName = Literal["qpf", "ptype"]
OperationType = Literal[
    "set_value",
    "increase",
    "decrease",
    "multiply",
    "clear",
    "ptype_set",
    "set_ptype",
    "screen_clear",
]


class EditParameters(BaseModel):
    value: float | None = None
    delta_mm: float | None = None
    factor: float | None = None
    target_ptype: int | None = None
    only_nonzero: bool = False
    threshold: float | None = None
    smooth_sigma: float = 0


class EditPreviewRequest(BaseModel):
    session_id: str
    tool: ToolName
    variable: VariableName
    operation: OperationType
    mask: dict[str, Any]
    parameters: EditParameters = Field(default_factory=EditParameters)

    @model_validator(mode="after")
    def validate_static_parameter_rules(self) -> Self:
        if (
            self.operation in {"clear", "decrease", "screen_clear"}
            and self.parameters.target_ptype is not None
        ):
            raise ValueError(
                f"target_ptype must not be set when operation={self.operation}"
            )
        return self


class EditPreviewResponse(BaseModel):
    preview_id: str
    affected_grid_count: int
    affected_area_km2: float
    area_mode: str = "spherical_lonlat_cell_area"
    before_stats: dict[str, Any]
    after_stats: dict[str, Any]
    op_ptype_transition: dict[str, int] | None = None
    new_precip_needs_ptype: bool = False
    new_precip_count: int = 0
    warnings: list[dict[str, Any]] = Field(default_factory=list)
    preview_image: str | None = None


class EditApplyRequest(BaseModel):
    session_id: str
    preview_id: str
    target_ptype: int | None = None


class EditApplyResponse(BaseModel):
    operation_id: str
    sequence_no: int
    applied: bool
    can_undo: bool
    can_redo: bool


class UndoRedoRequest(BaseModel):
    session_id: str


class UndoRedoResponse(BaseModel):
    can_undo: bool
    can_redo: bool
    operation_count: int


class OperationItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sequence_no: int
    tool_name: str
    operation_type: str
    variable_name: str
    affected_grid_count: int
    is_undone: int
    created_at: datetime


class OperationListResponse(BaseModel):
    operations: list[OperationItem]
